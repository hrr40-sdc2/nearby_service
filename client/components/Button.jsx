import React from 'react';
import styled from 'styled-components';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faChevronLeft, faChevronRight } from '@fortawesome/free-solid-svg-icons';

const ButtonStyle = styled.div`
  display: table-cell;
  vertical-align: middle;
  text-align: center;
`;

const Button = (props) => (
  <ButtonStyle className='button' value={props.value}> {props.value === 'right'
    ? <FontAwesomeIcon onClick={() => props.buttonClickHandler('right')} id='right' icon={faChevronLeft} size='2x'/>
    : <FontAwesomeIcon onClick={() => props.buttonClickHandler('left')} icon={faChevronRight} size='2x'/>}
  </ButtonStyle>
);

export default Button;
