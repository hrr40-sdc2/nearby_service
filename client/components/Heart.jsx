import React from 'react';
import styled from 'styled-components';

import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faHeart } from '@fortawesome/free-solid-svg-icons';
import { faHeart as farHeart } from '@fortawesome/free-regular-svg-icons';

const HeartButton = styled.div`
  position: absolute;
  top: 5%;
  right: 0;
  height: 20%;
  width: 15%;
  background-color: transparent;
  color: white;
  border: none;
  float: right;
`;

class Heart extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      clicked: false,
    };
  }

  checkArrForRender() {
    if (this.props.heartArr.indexOf(this.props.arrIndex) === -1) {
      this.setState({
        clicked: false,
      });
    } else {
      this.setState({
        clicked: true,
      });
    }
  }

  componentDidMount() {
    this.checkArrForRender();
  }

  heartClick() {
    console.log(this.props.arrIndex);
    if (this.props.heartArr.indexOf(this.props.arrIndex) === -1) {
      this.setState({
        clicked: true,
      });
    } else {
      this.setState({
        clicked: false,
      });
    }
    this.props.heartHouseClicked(this.props.arrIndex);
  }

  render() {
    return (
      <HeartButton>{this.state.clicked
        ? <FontAwesomeIcon
          id='fullHeart'
          icon={faHeart} size='2x'
          onClick={this.heartClick.bind(this)} />
        : <FontAwesomeIcon
          id='emptyHeart'
          icon={farHeart}
          size='2x'
          onClick={this.heartClick.bind(this)} />}
      </HeartButton>
    );
  }
}

export default Heart;
