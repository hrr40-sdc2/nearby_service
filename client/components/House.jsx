import styled from 'styled-components';

import React from 'react';
import Stars from './Stars.jsx';
import Heart from './Heart.jsx';

const ImageHouse = styled.img`
  width: 300px;
  height: 225px;
`;
const ImageContainer = styled.div`
  width: 300px;
  height: 225px;
  position: relative;
`;
const HouseContainer = styled.div`
  float: left;
  margin: 8px;
  width: 300px;
  height: 275px;
`;
const Words = styled.div`
  font-weight: 300;
  padding: 0;
  margin: 0;
`;
const LightDescription = styled.div`
  line-height: 30%;
  color: #616A6B;
  font-size: 13px;
`;

class House extends React.Component {
  handleClick() {
    this.props.changeCurrentHouse(this.props.house.id);
  }

  render() {
    const preUrl = 'https://hrr40-sdc2-jp.s3.us-east-2.amazonaws.com/';
    return (
      <HouseContainer className='house'>
        <ImageContainer>
          <Heart heartArr={this.props.heartArr} arrIndex={this.props.arrIndex}
            heartHouseClicked={this.props.heartHouseClicked} />
          <ImageHouse
            className='image'
            onClick={this.handleClick.bind(this)}
            src={preUrl + this.props.house.imgurl}>
          </ImageHouse>
        </ImageContainer>
        <Words onClick={this.handleClick.bind(this)}>
          <LightDescription className='layout'>
            <h5>
              {this.props.house.type.toUpperCase()}: {this.props.house.location.toUpperCase()}
            </h5>
          </LightDescription>
          <div className='description'>
            <h3>{this.props.house.title} </h3>
          </div>
          <LightDescription className='cost'>
            <h5>{`$${this.props.house.cost}/night`} </h5>
          </LightDescription>
          <div className='stars'>
            <h5>
              <Stars
                key={this.props.house.id}
                rating={this.props.house.stars}
              />
              {` ${this.props.house.reviewcount}`}
            </h5>
          </div>
        </Words>
      </HouseContainer>
    );
  }
}

export default House;
