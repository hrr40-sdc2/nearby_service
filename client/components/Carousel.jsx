import React from 'react';
import styled from 'styled-components';
import House from './House.jsx';
import Button from './Button.jsx';

const CarouselContainer = styled.div`
  width: 1500px;
  height: 450px;
`;

const ButtonContainer = styled.div`
  float: left;
  width: 25px;
  height: 225px;
  display: table;
  text-align: center;
  justify-content: center;
`;


class Carousel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      startIndex: 0,
      endIndex: 3,
      heartArr: [],
    };
  }

  getDisplayHouses(start, end) {
    this.setState({
      displayHouses: this.props.nearbyHouseList.slice(start, end),
    });
  }

  heartHouseClicked(index) {
    const newHeartArr = this.state.heartArr;
    if (newHeartArr.indexOf(index) === -1) {
      newHeartArr.push(index);
    } else {
      const targetIndex = newHeartArr.indexOf(index);
      newHeartArr.splice(targetIndex, 1);
    }
    this.setState({
      heartArr: newHeartArr,
    });
  }

  componentDidMount() {
    this.getDisplayHouses(this.state.startIndex, this.state.endIndex);
  }

  buttonClickHandler(side) {
    this.shiftDisplay(side);
  }

  shiftDisplay(side) {
    if (side === 'right') {
      const newStart = this.state.startIndex - 1;
      const newEnd = this.state.endIndex - 1;
      this.setState({
        startIndex: newStart,
        endIndex: newEnd,
      });
      this.getDisplayHouses(newStart, newEnd);
    } else if (side === 'left') {
      const newStart = this.state.startIndex + 1;
      const newEnd = this.state.endIndex + 1;
      this.setState({
        startIndex: newStart,
        endIndex: newEnd,
      });
      this.getDisplayHouses(newStart, newEnd);
    }
  }

  render() {
    return (
      <CarouselContainer className='carousel'>
        <ButtonContainer className='buttonDiv'>
          {this.state.startIndex === 0
            ? <ButtonContainer></ButtonContainer>
            : <Button
              value='right'
              buttonClickHandler={this.buttonClickHandler.bind(this)}
            />}
        </ButtonContainer>
        <div className='houseDisplay'>
          {this.state.displayHouses ? this.state.displayHouses.map((house) => (
            <House
              changeCurrentHouse={this.props.changeCurrentHouse}
              heartHouseClicked={this.heartHouseClicked.bind(this)}
              heartArr={this.state.heartArr}
              house={house}
              key={house.arrIndex}
            />
          ))
            : null}
        </div>
        <ButtonContainer className='buttonDiv'>
          {this.state.endIndex === 12
            ? <ButtonContainer></ButtonContainer>
            : <Button
              value='left'
              buttonClickHandler={this.buttonClickHandler.bind(this)}
            />}
        </ButtonContainer>
      </CarouselContainer>
    );
  }
}

export default Carousel;
