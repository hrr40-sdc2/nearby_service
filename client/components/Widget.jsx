import React from 'react';
import styled from 'styled-components';
import axios from 'axios';

import Carousel from './Carousel.jsx';

const WidgetContainer = styled.div`
  font-family: 'Tajawal', sans-serif;
  color: #424949;
  padding-left: 100px;
`;

class Widget extends React.Component {
  constructor() {
    super();
    this.state = {
      currentHouse: 1,
      view: false,
    };
  }

  changeCurrentHouse(houseId) {
    this.getNearbyHouses(houseId);
  }

  componentDidMount() {
    this.getNearbyHouses(11222);
  }

  updateHouseList(houses) {
    this.setState({
      nearbyHouseList: houses,
    });
  }

  getNearbyHouses(zip) {
    axios.get(`http://localhost:8081/houses/${zip}`)
      .then((houses) => {
        this.updateHouseList(houses.data);
        this.setState({
          view: true,
          currentHouse: houses.data[0].id,
        });
      })
      .catch((error) => {
        console.log(`error getting houses ${error}`);
      });
  }

  render() {
    return (
      <WidgetContainer id='morePlaces'>
        <h2>More places to stay</h2>
        {this.state.view
          ? <Carousel
            key={this.state.currentHouse}
            changeCurrentHouse={this.changeCurrentHouse.bind(this)}
            nearbyHouseList={this.state.nearbyHouseList} />
          : null}
      </WidgetContainer>
    );
  }
}

export default Widget;
