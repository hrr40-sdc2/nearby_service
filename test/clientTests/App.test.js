import { shallow, mount } from 'enzyme';
import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import Widget from '../../client/components/Widget.jsx';
import House from '../../client/components/House.jsx';
import Button from '../../client/components/Button.jsx';
import Carousel from '../../client/components/Carousel.jsx';
import Stars from '../../client/components/Stars.jsx';
import Heart from '../../client/components/Heart.jsx';

// Widget Component
describe('Test Widget component functionality', () => {
  const options = {
    disableLifecycleMethods: true,
  };

  it('should render correctly', () => {
    const component = shallow(<Widget/>, options);
    expect(component).toMatchSnapshot();
  });

  it('should call componentDidMount once', () => {
    const component = shallow(<Widget/>, options);
    component.instance().getNearbyHouses = jest.fn();
    component.update();
    const instance = component.instance();
    instance.componentDidMount();

    expect(instance.getNearbyHouses).toHaveBeenCalledTimes(1);
  });

  it('should render Carousel if state.view is true', () => {
    const component = shallow(<Widget />, options);
    component
      .setState({ view: true })
      .update();
    expect(component.find(Carousel)).toHaveLength(1);
  });

  // it('should change currentHouse state to the id changeCurrentHouse is called with', () => {
  //   const component = shallow(<Widget />, options);
  //   component.instance().getNearbyHouses = jest.fn();
  //   component
  //     .update()
  //     .instance()
  //     .changeCurrentHouse(4);
  //   expect(component.state('currentHouse')).toBe(4);
  // });

  it('should change update nearbyHouseList state to array of houses', () => {
    const houses = ['house1', 'house2', 'house3'];
    const component = shallow(<Widget />, options);
    component
      .instance()
      .updateHouseList(houses);
    expect(component.state('nearbyHouseList')).toBe(houses);
  });
});

// House Component testing
describe('Test House component functionality', () => {
  const clickFn = jest.fn();
  const props = {
    imgUrl: 'https://housemania.s3-us-west-1.amazonaws.com/annie-spratt-BmjlyHwV1S0-unsplash.jpg',
    cost: '$309/night',
    location: 'Saratoga',
    type: 'apartment',
    title: 'quaint house',
    stars: 4,
    reviewCount: 20,
  };

  it('should render correctly', () => {
    const component = shallow(<House house={props}/>);
    expect(component).toMatchSnapshot();
  });

  it('house button click call handleClick to run changeCurrentHouse', () => {
    const component = shallow(<House changeCurrentHouse={clickFn} house={props}/>);
    component
      .find('.image')
      .simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});

// Test Carousel Component
describe('Test Carousel component functionality', () => {
  const options = {
    disableLifecycleMethods: true,
  };
  const clickFn = jest.fn();

  it('should render correctly', () => {
    const component = shallow(<Carousel />, options);
    expect(component).toMatchSnapshot();
  });

  it('renders <House /> one times', () => {
    const nearbyHouseList = [
      {
        cost: 336,
        id: 6891904,
        imgurl: 'sdcimg-203.jpg',
        location: 'Favianfurt',
        reviewcount: 127,
        stars: 3,
        title: 'Fugit Numquam Dicta',
        type: 'at',
        zip: 37600
      },
      {
        cost: 246,
        id: 6891905,
        imgurl: 'sdcimg-204.jpg',
        location: 'Harveychester',
        reviewcount: 120,
        stars: 5,
        title: 'Sequi Voluptatem Nihil',
        type: 'temporibus',
        zip: 15169
      }
    ];
    const component = shallow(<Carousel />, options);
    component.setState({
      displayHouses: [
        {
          id: 1,
          imgurl: 'https://housemania.s3-us-west-1.amazonaws.com/annie-spratt-BmjlyHwV1S0-unsplash.jpg',
          cost: '$309/night',
          location: 'Saratoga',
          type: 'apartment',
          title: 'quaint house',
          stars: 4,
          reviewcount: 20,
        },
      ],
    });
    expect(component.find(House)).toHaveLength(1);
  });

  it('renders <Button/> two times for normal display', () => {
    const component = shallow(<Carousel />, options);
    component.setState({
      endIndex: 5,
      startIndex: 2,
    });
    expect(component.find('.buttonDiv')).toHaveLength(2);
  });


  it('should increase state.startIndex by 1 when shiftDisplay called with left', () => {
    const component = shallow(<Carousel />, options);
    component.instance().getDisplayHouses = clickFn;
    component.update();
    expect(component.state('startIndex')).toBe(0);
    component
      .instance()
      .shiftDisplay('left');
    expect(component.state('startIndex')).toBe(1);
  });

  it('should decrease state.startIndex by 1 when shiftDisplay called with right', () => {
    const component = shallow(<Carousel />, options);
    component.instance().getDisplayHouses = clickFn;
    component.setState({
      startIndex: 2,
      endIndex: 5,
    });
    component.update();
    expect(component.state('startIndex')).toBe(2);
    component
      .instance()
      .shiftDisplay('right');
    expect(component.state('startIndex')).toBe(1);
  });

  it('should call shiftDisplay function with a side', () => {
    const component = shallow(<Carousel />, options);
    component.instance().shiftDisplay = clickFn;
    component
      .update()
      .instance()
      .shiftDisplay('left');
    expect(component.instance().shiftDisplay).toHaveBeenCalledWith('left');
  });

  it('should call getDisplayHouses function with a start and end index', () => {
    const component = shallow(<Carousel />, options);
    component.instance().getDisplayHouses = clickFn;
    component
      .update()
      .instance()
      .getDisplayHouses(1, 3);
    expect(component.instance().getDisplayHouses).toHaveBeenCalledWith(1, 3);
  });

  it('should call componentDidMount once', () => {
    const component = shallow(<Carousel/>, options);
    component.instance().getDisplayHouses = jest.fn();
    component.update();
    const instance = component.instance();
    instance.componentDidMount();

    expect(instance.getDisplayHouses).toHaveBeenCalledTimes(1);
  });

  it('should add index to heartArr if not on it already', () => {
    const component = shallow(<Carousel />, options);
    component
      .setState({
        heartArr: [2],
      })
      .instance()
      .heartHouseClicked(5);
    expect(component.state('heartArr')).toMatchObject([2, 5]);
  });
});

// Test Button Component
describe('Button component functionality', () => {
  const clickFn = jest.fn();
  const props = {
    buttonClickHandler: jest.fn(),
    value: 'right',
  };

  it('should render correctly', () => {
    const component = shallow(<Button />);
    expect(component).toMatchSnapshot();
  });

  it('button value should be right', () => {
    const component = shallow(<Button value={'right'}/>);
    expect(component.prop('value')).toEqual('right');
  });


  it('button click call handleClick to run shift the display', () => {
    const component = shallow(<Button buttonClickHandler={clickFn}/>);
    component
      .find(FontAwesomeIcon)
      .simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});

// Test Stars Component
describe('Stars component functionality', () => {
  it('should render correctly', () => {
    const component = shallow(<Stars />);
    expect(component).toMatchSnapshot();
  });

  it('renders <FontAwesomIcon /> five times for every rating', () => {
    const component = shallow(<Stars />);
    component.setProps({
      rating: 4,
    });
    expect(component.find(FontAwesomeIcon)).toHaveLength(5);
  });
});

// Test Heart Component
describe('Heart component functionality', () => {
  const options = {
    disableLifecycleMethods: true,
  };

  it('should render correctly', () => {
    const component = shallow(<Heart />, options);
    expect(component).toMatchSnapshot();
  });

  it('should call componentDidMount once', () => {
    const component = shallow(<Heart/>, options);
    component.instance().checkArrForRender = jest.fn();
    component.update();
    const instance = component.instance();
    instance.componentDidMount();

    expect(instance.checkArrForRender).toHaveBeenCalledTimes(1);
  });

  it('should change clicked state to false if not in heartArr with checkArrForRender', () => {
    const component = shallow(<Heart />, options);
    component
      .setProps({
        heartArr: [1],
        arrIndex: 2,
      })
      .instance()
      .checkArrForRender();
    expect(component.state('clicked')).toBe(false);
  });

  it('should toggle state clicked when the heart is clicked', () => {
    const component = shallow(<Heart />, options);
    component
      .setProps({
        heartArr: [1],
        arrIndex: 2,
        heartHouseClicked: jest.fn(),
      })
      .find('#emptyHeart')
      .simulate('click');
    expect(component.state('clicked')).toBe(true);
  });
});
