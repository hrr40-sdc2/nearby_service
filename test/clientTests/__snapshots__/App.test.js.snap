// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component functionality should render correctly 1`] = `
<styled.div
  className="button"
>
   
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          320,
          512,
          Array [],
          "f054",
          "M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z",
        ],
        "iconName": "chevron-right",
        "prefix": "fas",
      }
    }
    inverse={false}
    listItem={false}
    mask={null}
    onClick={[Function]}
    pull={null}
    pulse={false}
    rotation={null}
    size="2x"
    spin={false}
    symbol={false}
    title=""
    transform={null}
  />
</styled.div>
`;

exports[`Heart component functionality should render correctly 1`] = `
<styled.div>
  <FontAwesomeIcon
    border={false}
    className=""
    fixedWidth={false}
    flip={null}
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f004",
          "M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z",
        ],
        "iconName": "heart",
        "prefix": "far",
      }
    }
    id="emptyHeart"
    inverse={false}
    listItem={false}
    mask={null}
    onClick={[Function]}
    pull={null}
    pulse={false}
    rotation={null}
    size="2x"
    spin={false}
    symbol={false}
    title=""
    transform={null}
  />
</styled.div>
`;

exports[`Stars component functionality should render correctly 1`] = `
<styled.div
  className="star"
/>
`;

exports[`Test Carousel component functionality should render correctly 1`] = `
<styled.div
  className="carousel"
>
  <styled.div
    className="buttonDiv"
  >
    <styled.div />
  </styled.div>
  <div
    className="houseDisplay"
  />
  <styled.div
    className="buttonDiv"
  >
    <Button
      buttonClickHandler={[Function]}
      value="left"
    />
  </styled.div>
</styled.div>
`;

exports[`Test House component functionality should render correctly 1`] = `
<styled.div
  className="house"
>
  <styled.div>
    <Heart />
    <styled.img
      className="image"
      onClick={[Function]}
      src="https://hrr40-sdc2-jp.s3.us-east-2.amazonaws.com/undefined"
    />
  </styled.div>
  <styled.div
    onClick={[Function]}
  >
    <styled.div
      className="layout"
    >
      <h5>
        APARTMENT
        : 
        SARATOGA
      </h5>
    </styled.div>
    <div
      className="description"
    >
      <h3>
        quaint house
         
      </h3>
    </div>
    <styled.div
      className="cost"
    >
      <h5>
        $$309/night/night
         
      </h5>
    </styled.div>
    <div
      className="stars"
    >
      <h5>
        <Stars
          rating={4}
        />
         undefined
      </h5>
    </div>
  </styled.div>
</styled.div>
`;

exports[`Test Widget component functionality should render correctly 1`] = `
<styled.div
  id="morePlaces"
>
  <h2>
    More places to stay
  </h2>
</styled.div>
`;
